librecrypt_crypt.3
.TH LIBRECRYPT_CRYPT 3 LIBRECRYPT
.SH NAME
librecrypt_crypt - Compute password hash encoded in ASCII with settings prefix

.SH SYNOPSIS
.nf
#include <librecrypt.h>

ssize_t \fBlibrecrypt_crypt\fP(char *restrict \fIout_buffer\fP, size_t \fIsize\fP,
                         const char *\fIphrase\fP, size_t \fIlen\fP,
                         const char *\fIsettings\fP, LIBRECRYPT_CONTEXT *\fIctx\fP);
.fi
.PP
Link with
.IR -lrecrypt .
Static linking may require additional flags
depending on enabled hash algorithms.

.SH DESCRIPTION
The
.BR librecrypt_crypt ()
function computes the hash of a password.
The resulting hash is encoded with
.BR librecrypt_encode (3)
and the settings string is included in the front
of the output.
.PP
The password is provided in
.IR phrase
and can contain null bytes; its length is
specified, in bytes, by
.IR len .
.PP
The password hash configuration string is
provided in
.IR settings .
If
.I settings
contains a resulting hash, it is ignored.
.PP
The
.I ctx
parameter is used for providing library
configurations, see
.BR librecrypt_create_context (3)
for more information; used
.I NULL
for default configurations.
.PP
On successful completion, if
.I size
is positive, the output is
null byte-terminated even if truncated.
.PP
Any encountered
.BR EINTR
is ignored.
.PP
On failure,
.I out_buffer
may be partially written.
.PP
.I settings
must not be
.IR NULL .

.SH RETURN VALUES
The
.BR librecrypt_crypt ()
function returns the number of bytes that would
have been written to
.IR out_buffer
if
.I size
was sufficiently large, excluding the terminating
null byte. On failure, -1 is returned and
.I errno
is set to describe the error.

.SH ERRORS
The
.BR librecrypt_crypt ()
function will fail if:
.TP
.B EINVAL
.I settings
is invalid.
.TP
.B ERANGE
.I len
is too large or too small for the selected
initial algorithm.
.TP
.B EOVERFLOW
The expected return value is greater than {SSIZE_MAX}.
.TP
.B ENOMEM
Failed to allocate internal scratch memory.
.TP
.B ENOSYS
A selected hash algorithm is not recognised
or was disabled at compile-time.

.SH ATTRIBUTES
For an explanation of the terms used in this section, see
.BR attributes (7).
.PP
.TS
allbox;
lb lb lb
l l l.
Interface	Attribute	Value
T{
.BR librecrypt_crypt ()
T}	Thread safety	MT-Safe
T{
.BR librecrypt_crypt ()
T}	Async-signal safety	AS-Unsafe
.TE
.sp

.SH HISTORY
The
.BR librecrypt_crypt ()
function was introduced in version 1.0 of
.BR librecrypt .

Since version 1.1 of the
.BR librecrypt_crypt ()
function, the
.I ctx
parameter (previously called
.IR reserved )
has the type
.B LIBRECRYPT_CONTEXT *
(previously
.BR "void *" ),
and setting it to
.RI non- NULL
no longer causes failure with
.I errno
set to
.IR EINVAL .

.SH SEE ALSO
.BR librecrypt (7),
.BR librecrypt_hash_binary (3),
.BR librecrypt_hash (3),
.BR librecrypt_verify (3),
.BR librecrypt_test_supported (3)


librecrypt_crypt.c
/* See LICENSE file for copyright and license details. */
#include "common.h"
#ifndef TEST


ssize_t
librecrypt_crypt(char *restrict out_buffer, size_t size, const char *phrase,
                 size_t len, const char *settings, LIBRECRYPT_CONTEXT *ctx)
{
	return librecrypt_hash_(out_buffer, size, phrase, len, settings, ctx, ASCII_CRYPT);
}


#else


#define SP4 "    "
#define SP20 SP4 SP4 SP4 SP4 SP4
#define SP80 SP20 SP20 SP20 SP20
#define SP84 SP80 SP4

#define GET_ARGON2_SCRATCH_SIZE(HASHLEN) ((HASHLEN) > 64u ? ((HASHLEN) + 63u) & ~31u : (HASHLEN))


static void
check(const char *phrase, const char *settings, const char *chain, size_t chain_prefix, const char *hash,
      size_t hash_prefix, size_t scratchsize, LIBRECRYPT_CONTEXT *ctx)
{
	size_t hashlen = strlen(hash);
	size_t len = strlen(phrase);
	char buf[1024], buf2[sizeof(buf)], expected[sizeof(buf)], pad;
	int strict_pad;
	const void *lut;
	ssize_t r;

	assert(hashlen <= sizeof(buf));

	CANARY_FILL(buf);
	EXPECT(librecrypt_crypt(buf, sizeof(buf), phrase, len, settings, ctx) == (ssize_t)hashlen);
	EXPECT(!memcmp(hash, buf, hashlen + 1u));
	CANARY_X_CHECK(buf, hashlen + 1u, scratchsize);

	CANARY_FILL(buf);
	EXPECT(librecrypt_crypt(buf, hashlen + 1u, phrase, len, settings, ctx) == (ssize_t)hashlen);
	EXPECT(!memcmp(hash, buf, hashlen + 1u));
	CANARY_X_CHECK(buf, hashlen + 1u, scratchsize);

	CANARY_FILL(buf);
	EXPECT(librecrypt_crypt(buf, hashlen, phrase, len, settings, ctx) == (ssize_t)hashlen);
	EXPECT(!memcmp(hash, buf, hashlen - 1u));
	EXPECT(!buf[hashlen - 1u]);
	CANARY_X_CHECK(buf, hashlen, scratchsize);

	CANARY_FILL(buf);
	EXPECT(librecrypt_crypt(buf, 2u, phrase, len, settings, ctx) == (ssize_t)hashlen);
	EXPECT(!memcmp(hash, buf, 1u));
	EXPECT(!buf[1u]);
	CANARY_X_CHECK(buf, 2u, 2u);

	CANARY_FILL(buf);
	EXPECT(librecrypt_crypt(buf, 1u, phrase, len, settings, ctx) == (ssize_t)hashlen);
	EXPECT(!buf[0u]);
	CANARY_X_CHECK(buf, 1u, 1u);

	EXPECT(librecrypt_crypt(buf, 0u, phrase, len, settings, ctx) == (ssize_t)hashlen);
	EXPECT(librecrypt_crypt(NULL, 0u, phrase, len, settings, ctx) == (ssize_t)hashlen);

	lut = librecrypt_get_encoding(settings, strlen(settings), &pad, &strict_pad, 1, ctx);
	assert(lut);
	r = librecrypt_decode(expected, sizeof(expected), &hash[hash_prefix], hashlen - hash_prefix, lut, pad, strict_pad);
	assert(r > 0 && (size_t)r <= sizeof(expected));

	CANARY_FILL(buf);
	CANARY_FILL(buf2);
	EXPECT(librecrypt_crypt(buf, sizeof(buf), expected, (size_t)r, settings, ctx) == (ssize_t)hashlen);
	errno = 0;
	EXPECT(librecrypt_crypt(buf2, sizeof(buf2), phrase, len, chain, ctx) == (ssize_t)(hashlen - hash_prefix + chain_prefix));
	EXPECT(!memcmp(buf2, chain, chain_prefix));
	EXPECT(!memcmp(&buf[hash_prefix], &buf2[chain_prefix], hashlen - hash_prefix + 1u));
	CANARY_X_CHECK(buf, hashlen, scratchsize);
	CANARY_X_CHECK(buf2, hashlen - hash_prefix + chain_prefix, scratchsize);
}


#define CHECK(PHRASE, CONF, HASHLEN, IS_DEFAULT_HASHLEN, HASH)\
	do {\
		size_t scratchsize = GET_SCRATCH_SIZE(HASHLEN);\
		check(PHRASE, CONF HASH, CONF "*" #HASHLEN ">" CONF HASH,\
		      sizeof(CONF "*" #HASHLEN ">" CONF) - 1u, CONF HASH,\
		      sizeof(CONF) - 1u, scratchsize, ctx);\
		check(PHRASE, CONF "*" #HASHLEN, CONF "*" #HASHLEN ">" CONF "*" #HASHLEN,\
		      sizeof(CONF "*" #HASHLEN ">" CONF) - 1u, CONF HASH,\
		      sizeof(CONF) - 1u, scratchsize, ctx);\
		if (IS_DEFAULT_HASHLEN) {\
			check(PHRASE, CONF, CONF ">" CONF, sizeof(CONF ">" CONF) - 1u,\
			      CONF HASH, sizeof(CONF) - 1u, scratchsize, ctx);\
		}\
	} while (0)


#define CHECK_BAD(ALGO)\
	do {\
		errno = 0;\
		EXPECT(librecrypt_crypt(NULL, 0u, NULL, 0u, ALGO"m=0,t=999999999999999999,p=0$AAAABBBB$*0", ctx) == -1);\
		EXPECT(errno == EINVAL);\
	} while (0)


int
main(void)
{
	char buf[1024], buf2[1024], conf[256], nuls[256];
	LIBRECRYPT_CONTEXT *ctx = NULL;
	ssize_t r;

	SET_UP_ALARM();
	INIT_RESOURCE_TEST();

#if defined(__linux__)
	libtest_getrandom_real = 0;
	libtest_getrandom_error = ENOSYS;
#endif

#define GET_SCRATCH_SIZE(HASHLEN) GET_ARGON2_SCRATCH_SIZE(HASHLEN)
#if defined(SUPPORT_ARGON2I)
	r = snprintf(conf, sizeof(conf), "$argon2i$m=256,t=8,p=1$AAAABBBBCCCC$*%zu", SIZE_MAX / 4u * 3u + 3u);
	assert(r > 0 && (size_t)r < sizeof(conf));
	errno = 0;
	EXPECT(librecrypt_crypt(NULL, 0u, NULL, 0u, conf, ctx) == -1);
# if SIZE_MAX > UINT32_MAX
	EXPECT(errno == EINVAL);
# else
	EXPECT(errno == EOVERFLOW);
	if (libtest_have_custom_malloc()) {
		libtest_pretend_allocation_successful = 1;
		errno = 0;
		EXPECT(librecrypt_crypt(buf, sizeof(buf), NULL, 0u, conf, ctx) == -1);
		libtest_pretend_allocation_successful = 0;
		EXPECT(errno == EOVERFLOW);
	}
# endif

# if SIZE_MAX == UINT32_MAX
	r = snprintf(conf, sizeof(conf), "$argon2i$m=256,t=8,p=1$AAAABBBBCCCC$*%zu", (SIZE_MAX / 4u * 3u) / 2u);
	assert(r > 0 && (size_t)r < sizeof(conf));
	errno = 0;
	EXPECT(librecrypt_crypt(NULL, 0u, NULL, 0u, conf, ctx) == -1);
	EXPECT(errno == EOVERFLOW);
# endif

# if SIZE_MAX == UINT32_MAX
	r = snprintf(conf, sizeof(conf), "$argon2i$m=256,t=8,p=1$AAAABBBBCCCC$*%zu", SIZE_MAX / 4u * 3u);
	assert(r > 0 && (size_t)r < sizeof(conf));
	errno = 0;
	EXPECT(librecrypt_crypt(NULL, 0u, NULL, 0u, conf, ctx) == -1);
	EXPECT(errno == EOVERFLOW);
# endif

	CHECK("password",  "$argon2i$"   "m=256,t=2,p=1$c29tZXNhbHQ$",  32, 1, "/U3YPXYsSb3q9XxHvc0MLxur+GP960kN9j7emXX8zwY");
	CHECK("password",  "$argon2i$v=19$m=256,t=2,p=1$c29tZXNhbHQ$",  32, 1, "iekCn0Y3spW+sCcFanM2xBT63UP2sghkUoHLIUpWRS8");
	CHECK_BAD("$argon2i$");
#endif
#if defined(SUPPORT_ARGON2ID)
	CHECK("password", "$argon2id$v=19$m=256,t=2,p=1$c29tZXNhbHQ$",  32, 1, "nf65EOgLrQMR/uIPnA4rEsF5h7TKyQwu9U1bMCHGi/4");
	CHECK_BAD("$argon2id$");
#endif
#if defined(SUPPORT_ARGON2DS)
	CHECK("",         "$argon2ds$v=16$m=""8,t=1,p=1$ICAgICAgICA$",  32, 1, "zgdykk9ZjN5VyrW0LxGw8LmrJ1Z6fqSC+3jPQtn4n0s");
	CHECK_BAD("$argon2ds$");
#endif
#if defined(SUPPORT_ARGON2D)
	CHECK("",          "$argon2d$v=16$m=""8,t=1,p=1$ICAgICAgICA$", 100, 0, "NjODMrWrS7zeivNNpHsuxD9c6uDmUQ6YqPRhb8H5DSNw9"
	                                                                       "n683FUCJZ3tyxgfJpYYANI+01WT/S5zp1UVs+qNRwnkdE"
	                                                                       "yLKZMg+DIOXVc9z1po9ZlZG8+Gp4g5brqfza3lvkR9vw");
	CHECK_BAD("$argon2d$");
#endif
#if defined(SUPPORT_ARGON2ID)
	assert(!libtest_getentropy_error);

	libtest_getentropy_real = 0;
	libtest_random_pattern = (const unsigned char *)"\x00\x01\x02\x03";
	/* since librecrypt_realise_salts doesn't generate random data then base64-encode it,
	 * but rather just takes random characters from base64 alphabet (with restrictions
	 * one the list one if the count isn't a multiple of 4), this will map to a repeation
	 * of "ABCD", rather the become "AAECAwABAgMAAQIDAAECAwAB" */
	libtest_random_pattern_length = 4u;
	libtest_random_pattern_offset = 0u;
	CANARY_FILL(buf);
	r = librecrypt_crypt(buf, sizeof(buf), "", 0u, "$argon2id$v=19$m=8,t=1,p=1$*18$*33", ctx);
	libtest_random_pattern = NULL;
	libtest_random_pattern_length = 0u;
	libtest_random_pattern_offset = 0u;
	libtest_getentropy_real = 1;
	EXPECT(r > 0);
	assert((size_t)r < sizeof(buf));
	EXPECT((size_t)r == sizeof("$argon2id$v=19$m=8,t=1,p=1$$") - 1u + 24u + 44u);
	EXPECT(!buf[r]);
	CANARY_FILL(buf2);
	EXPECT(librecrypt_crypt(buf2, sizeof(buf2), "", 0u, buf, ctx) == r);
	EXPECT(!memcmp(buf, buf2, (size_t)r + 1u));
	EXPECT(!memcmp(buf, "$argon2id$v=19$m=8,t=1,p=1$ABCDABCDABCDABCDABCDABCD$",
	             sizeof("$argon2id$v=19$m=8,t=1,p=1$ABCDABCDABCDABCDABCDABCD$") - 1u));
	CANARY_X_CHECK(buf, (size_t)r + 1u, 33u);
	CANARY_X_CHECK(buf2, (size_t)r + 1u, 33u);

	libtest_getentropy_real = 0;
	libtest_random_pattern = (const unsigned char *)"\x00\x01\x02\03";
	libtest_random_pattern_length = 4u;
	libtest_random_pattern_offset = 0u;
	CANARY_FILL(buf);
	r = librecrypt_crypt(buf, sizeof(buf), "", 0u, "$argon2id$v=19$m=8,t=1,p=1$*18$*33>"
	                                               "$argon2id$v=19$m=8,t=1,p=1$*18$*33", ctx);
	libtest_random_pattern = NULL;
	libtest_random_pattern_length = 0u;
	libtest_random_pattern_offset = 0u;
	libtest_getentropy_real = 1;
	EXPECT(r > 0);
	assert((size_t)r < sizeof(buf));
	EXPECT((size_t)r == sizeof("$argon2id$v=19$m=8,t=1,p=1$$*33>$argon2id$v=19$m=8,t=1,p=1$$") - 1u + 2u * 24u + 44u);
	EXPECT(!buf[r]);
	CANARY_FILL(buf2);
	EXPECT(librecrypt_crypt(buf2, sizeof(buf2), "", 0u, buf, ctx) == r);
	EXPECT(!memcmp(buf, buf2, (size_t)r + 1u));
	EXPECT(!memcmp(buf, "$argon2id$v=19$m=8,t=1,p=1$ABCDABCDABCDABCDABCDABCD$*33>"
	                    "$argon2id$v=19$m=8,t=1,p=1$ABCDABCDABCDABCDABCDABCD$",
	             sizeof("$argon2id$v=19$m=8,t=1,p=1$ABCDABCDABCDABCDABCDABCD$*33>"
	                    "$argon2id$v=19$m=8,t=1,p=1$ABCDABCDABCDABCDABCDABCD$") - 1u));
	CANARY_X_CHECK(buf, (size_t)r + 1u, 33u);
	CANARY_X_CHECK(buf2, (size_t)r + 1u, 33u);
#endif
#undef GET_SCRATCH_SIZE

#if defined(__linux__)
	libtest_getrandom_real = 1;
	libtest_getrandom_error = 0;
#endif

	ctx = librecrypt_create_context();
	assert(ctx != NULL);
	memset(nuls, 0, sizeof(nuls));

#if defined(SUPPORT_ARGON2I)
# define GET_SCRATCH_SIZE(HASHLEN) GET_ARGON2_SCRATCH_SIZE(HASHLEN)
	assert(sizeof(nuls) >= 4u);
	assert(librecrypt_set_pepper(ctx, LIBRECRYPT_ARGON2I_V1_3, nuls, 4u) == 0);
	CHECK(" ",  "$argon2i$v=19$m=8,t=1,p=1$ICAgICAgICA$", 8, 0, "Mhl4o3AkJuA");
	CHECK(SP84, "$argon2i$v=19$m=8,t=1,p=1$ICAgICAgICA$", 8, 0, "+hlEcRn+F3s");
	CHECK(SP80, "$argon2i$v=19$m=8,t=1,p=1$ICAgICAgICA$", 8, 0, "z2d6ce8UqS0");

	assert(sizeof(nuls) >= 140u);
	assert(librecrypt_set_pepper(ctx, LIBRECRYPT_ARGON2I_V1_3, nuls, 140u) == 0);
	CHECK(SP80, "$argon2i$v=19$m=8,t=1,p=1$ICAgICAgICA$", 8, 0, "15FAGe1KIX8");

	assert(sizeof(nuls) >= 160u);
	assert(librecrypt_set_pepper(ctx, LIBRECRYPT_ARGON2I_V1_3, nuls, 160u) == 0);
	CHECK(SP80, "$argon2i$v=19$m=8,t=1,p=1$ICAgICAgICA$", 8, 0, "oH3H5atuca8");

	assert(sizeof(nuls) >= 128u);
	assert(librecrypt_set_pepper(ctx, LIBRECRYPT_ARGON2I_V1_3, nuls, 128u) == 0);
	CHECK(SP80, "$argon2i$v=19$m=8,t=1,p=1$ICAgICAgICA$", 8, 0, "TsimqI1YC08");

	assert(sizeof(nuls) >= 256u);
	assert(librecrypt_set_pepper(ctx, LIBRECRYPT_ARGON2I_V1_3, nuls, 256u) == 0);
	CHECK(SP80, "$argon2i$v=19$m=8,t=1,p=1$ICAgICAgICA$", 8, 0, "mzPlVOVjVos");
# undef GET_SCRATCH_SIZE
#endif

	librecrypt_free_context(ctx);

	STOP_RESOURCE_TEST();
	return 0;
}


#endif
